packages/mds-compliance-service/repository/index.ts

```diff
@@ -146,42 +146,61 @@ class ComplianceReadWriteRepository extends ReadWriteRepository {
     }
   }
 
+  // TODO add a beforeCommit fn param
   public createComplianceSnapshot = async (
-    complianceSnapshot: ComplianceSnapshotDomainModel
+    complianceSnapshot: ComplianceSnapshotDomainModel,
+    beforeCommit: (domain: ComplianceSnapshotDomainModel) => Promise<void>
   ): Promise<ComplianceSnapshotDomainModel> => {
     const { connect } = this
     try {
+      // is there a way to pass in beforeCommit here?
       const connection = await connect('rw')
-      const {
-        raw: [entity]
-      }: InsertReturning<ComplianceSnapshotEntity> = await connection
-        .getRepository(ComplianceSnapshotEntity)
-        .createQueryBuilder()
-        .insert()
-        .values([ComplianceSnapshotDomainToEntityCreate.map(complianceSnapshot)])
-        .returning('*')
-        .execute()
-      return ComplianceSnapshotEntityToDomain.map(entity)
+      return connection.transaction(async trans => {
+        const {
+          raw: [entity]
+        }: InsertReturning<ComplianceSnapshotEntity> = await trans
+          .getRepository(ComplianceSnapshotEntity)
+          .createQueryBuilder()
+          .insert()
+          .values([ComplianceSnapshotDomainToEntityCreate.map(complianceSnapshot)])
+          .returning('*')
+          .execute()
+        const domain = ComplianceSnapshotEntityToDomain.map(entity)
+
+        // submit to kafka or whatever - but if there's an exception, it blows up and aborts the commit
+        await beforeCommit(domain)
+
+        return domain
+      })
     } catch (error) {
       throw RepositoryError(error)
     }
   }
 
+  // TODO: add a beforeCommit fn param
   public createComplianceSnapshots = async (
-    ComplianceSnapshots: ComplianceSnapshotDomainModel[]
+    ComplianceSnapshots: ComplianceSnapshotDomainModel[],
+    beforeCommit: (domains: ComplianceSnapshotDomainModel[]) => Promise<void>
   ): Promise<ComplianceSnapshotDomainModel[]> => {
     if (ComplianceSnapshots.length === 0) return []
     const { connect } = this
     try {
       const connection = await connect('rw')
-      const { raw: entities }: InsertReturning<ComplianceSnapshotEntity> = await connection
-        .getRepository(ComplianceSnapshotEntity)
-        .createQueryBuilder()
-        .insert()
-        .values(ComplianceSnapshots.map(ComplianceSnapshotDomainToEntityCreate.mapper()))
-        .returning('*')
-        .execute()
-      return entities.map(ComplianceSnapshotEntityToDomain.map)
+      return connection.transaction(async trans => {
+        const { raw: entities }: InsertReturning<ComplianceSnapshotEntity> = await trans
+          .getRepository(ComplianceSnapshotEntity)
+          .createQueryBuilder()
+          .insert()
+          .values(ComplianceSnapshots.map(ComplianceSnapshotDomainToEntityCreate.mapper()))
+          .returning('*')
+          .execute()
+        const domains = entities.map(ComplianceSnapshotEntityToDomain.map)
```

Comment:
Typically we would use `ComplianceSnapshotEntityToDomain.mapper()` here (which creates the map function and allows options to be passed in. The net effect is the same, but that's the standard pattern.
```suggestion
        const domains = entities.map(ComplianceSnapshotEntityToDomain.mapper())
```